workspaces/arborist/lib/arborist/build-ideal-tree.js

```diff
@@ -1226,9 +1226,22 @@ This is a one-time fix-up, please be patient...
     const { installLinks, legacyPeerDeps } = this
     const isWorkspace = this.idealTree.workspaces && this.idealTree.workspaces.has(spec.name)
 
-    // spec is a directory, link it unless installLinks is set or it's a workspace
+    // spec is a directory, link it if:
+    // - it's a workspace, OR
+    // - it's a project-internal file: dependency (always linked), OR
+    // - it's external and installLinks is false
     // TODO post arborist refactor, will need to check for installStrategy=linked
-    if (spec.type === 'directory' && (isWorkspace || !installLinks)) {
+    let isProjectInternalFileSpec = false
+    if (edge && edge.rawSpec && edge.rawSpec.startsWith('file:') &&
+        (edge.rawSpec.startsWith('file:../') || edge.rawSpec.startsWith('file:./'))) {
+      const targetPath = resolve(parent.realpath, edge.rawSpec.replace(/^file:/, ''))
+      const resolvedProjectRoot = resolve(this.idealTree.realpath)
+      // Check if the target is within the project root
+      isProjectInternalFileSpec = targetPath.startsWith(resolvedProjectRoot + sep) || targetPath === resolvedProjectRoot
+    }
+    // Decide whether to link or copy the dependency
+    const shouldLink = isWorkspace || isProjectInternalFileSpec || (!isProjectInternalFileSpec && !installLinks)
```

Comment:
```suggestion
    const shouldLink = isWorkspace || isProjectInternalFileSpec || !installLinks
```

Boolean logic is fun!  If we get to the last `||` we already know that `!isWorkspace` and `!isProjectInternalFileSpec`